### PR TITLE
Implementing Superbuild Structure for CMaize Project

### PR DESCRIPTION
**Pull request by eminokic**

### CODE REVIEW

File: LICENSE [hunk 1]
```diff
@@ -1,201 +0,0 @@
-                                 Apache License
```
Comment (ryanmrichard) [inline LICENSE]: Please put the license back

File: Shared-Library/hello-world.hpp [hunk 2]
```diff
@@ -0,0 +1,5 @@
+#pragma once
+
+string say_hello();
+
+int main();
```
Comment (ryanmrichard) [inline Shared-Library/hello-world.hpp]: shared libraries don't have main functions, also there's no point in forward delcaring the main function since it defines the API to an executable and not a library.

**Review by ryanmrichard — CHANGES_REQUESTED** [review]
do not check in the build directory. I'd recommend adding your build directory names it to the `.gitignore` file so you don't accidentally check them in

File: CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,37 @@
+cmake_minimum_required(VERSION 3.5 FATAL_ERROR)
+
+project(CMaizeSuperbuild LANGUAGES CXX)
+
+list(APPEND BOOST_COMPONENTS_REQUIRED filesystem system)
```
Comment (ryanmrichard) [inline CMakeLists.txt]: For now I recommend having no dependencies in this project. Once your superbuild works for your hello world library then you can make the library depend on CMaizeExamples/CMakePublic. I wouldn't add any other dependencies to the project beyond that.

File: external/upstream/boost/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,104 @@
+find_package(Boost ${Boost_MINIMUM_REQUIRED} QUIET COMPONENTS "${BOOST_COMPONENTS_REQUIRED}")
```
Comment (ryanmrichard) [inline external/upstream/boost/CMakeLists.txt]: I could be wrong, but I'm guessing you copy/pasted this from somewhere. If that's the case you need to attribute the source. That said as my previous comment suggested you don't want to use Boost in this project.

File: src/hello-world.cpp [hunk 5]
```diff
@@ -0,0 +1,10 @@
+#include <cstdlib>
+#include <iostream>
+#include <string>
+
+std::string say_hello() { return std::string("Hello, CMake superbuild!"); }
+
+int main() {
```
Comment (ryanmrichard) [inline src/hello-world.cpp]: So two things. First, in C++ it's customary to put the definition of the main function in a file `main.cpp`. Second your CMakeLists.txt isn't building an executable (it's building a library) so there shouldn't be a main function (all hell's going to break loose when someone tries to link to this library because there can't be two main functions). We ultimately need a library. I'd avoid building an executable for now (we can build one later as a test, but it's going to add more complexity).

File: CMakeLists.txt [hunk 6]
```diff
@@ -2,6 +2,10 @@ cmake_minimum_required(VERSION 3.5 FATAL_ERROR)
 
 project(CMaizeSuperbuild LANGUAGES CXX)
 
+# Required to enable unit testing
+include(CTest)
+enable_testing()
```
Comment (ryanmrichard) [inline CMakeLists.txt]: don't explicitly call `enable_testing()`. `include(CTest)` will call it if `BUILD_TESTING` is true

File: CMakeLists.txt [hunk 7]
```diff
@@ -2,6 +2,10 @@ cmake_minimum_required(VERSION 3.5 FATAL_ERROR)
...
+
 set(CMAKE_CXX_STANDARD 11)
```
Comment (ryanmrichard) [inline CMakeLists.txt]: You aren't using an C++11 features, so don't force it. Anyways the C++ standard should be set on a target not globally. Also you should never call `set()` for variables defined by CMake (such as you do on lines 9 to 11); doing so will override the user's options. Usually you want to use the `option` command (which will only set them if the user hasn't).

File: CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    CMAKE_ARGS -DCMAKE_CXX_COMPILER="${CMAKE_CXX_COMPILER}"
+               -DCMAKE_CXX_STANDARD="${CMAKE_CXX_STANDARD}"
+               -DCMAKE_CXX_EXTENSIONS="${CMAKE_CXX_EXTENSIONS}"
+               -DCMAKE_CXX_STANDARD_REQUIRED="${CMAKE_CXX_STANDARD_REQUIRED}"
+    CMAKE_CACHE_ARGS -DCMAKE_CXX_FLAGS:STRING="${CMAKE_CXX_FLAGS}"
+    BUILD_ALWAYS 1
+    INSTALL_COMMAND ""
+)
+
+ExternalProject_Add(CMakePublic
```
Comment (ryanmrichard) [inline CMakeLists.txt]: Let's not worry about CMakePublic until your library builds and installs. Also `ExternalProject_Add` has a configure step so don't add one.

File: src/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+        )
+
+target_include_directories(${PROJECT_NAME} PUBLIC
+        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include/>
+        $<INSTALL_INTERFACE:include/>
+        )
+
+install(TARGETS ${PROJECT_NAME}
+        EXPORT ${PROJECT_NAME}-targets
+        PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_INCLUDEDIR}
```
Comment (ryanmrichard) [inline src/CMakeLists.txt]: Pretty sure you need `include(GNUInstallDirs)` before you can use `${CMAKE_INSTALL_INCLUDEDIR}`

**Review by ryanmrichard — CHANGES_REQUESTED** [review]

File: CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,35 @@
+cmake_minimum_required(VERSION 3.0 FATAL_ERROR)
+
+project(CMaizeSuperbuild)
+
+include(CTest)
+include(ExternalProject)
+include(FetchContent)
```
Comment (ryanmrichard) [inline CMakeLists.txt]: You don't need `FetchContent`. `ExternalProject_Add` can get souce code. As a disclaimer I think way down the road we'll actually want to use `FetchContent` to get the source code (it'll make it easier to build offline, and for the user to override the default dependency version), but let's not worry about that until we have the superbuild in place.

File: CMake/Depends/CMakePublic.cmake [hunk 11]
```diff
@@ -0,0 +1,19 @@
+if(NOT TARGET depends::CMakePublic)
```
Comment (ryanmrichard) [inline CMake/Depends/CMakePublic.cmake]: Don't worry about CMakePublic for this PR. Let's make this PR about building and installing your hello world library via external project add. I'd recommend just deleting all of the CMakePublic stuff you added; it's not what we'll need down the road either.

File: CMake/Modules/SuperBuildDepends.cmake [hunk 12]
```diff
@@ -0,0 +1,62 @@
+# MIT License
```
Comment (ryanmrichard) [inline CMake/Modules/SuperBuildDepends.cmake]: For the purposes of this repo you shouldn't be using other people's work, so remove this file. This repo is going to be a textbook superbuild example with no frills.

File: CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+set_property(DIRECTORY PROPERTY EP_BASE ${CMAKE_CURRENT_BINARY_DIR}/subprojects)
+
+set(STAGED_INSTALL_PREFIX ${CMAKE_BINARY_DIR}/stage)
+message(STATUS "${PROJECT_NAME} staged install: ${STAGED_INSTALL_PREFIX}")
+
+
+include_directories(${CMAKE_CURRENT_SOURCE_DIR}/src)
+include_directories(${EXTERNAL_INSTALL_LOCATION}/Shared-Library)
+
+add_library(hello-world)
```
Comment (ryanmrichard) [inline CMakeLists.txt]: This needs to go back to how you had it. I'm not sure why you pulled it up into the driver `CMakeLists.txt`.

File: Shared-Library/hello-world.hpp [hunk 14]
```diff
@@ -0,0 +1,5 @@
+#pragma once
+#include <core/version.h>
```
Comment (ryanmrichard) [inline Shared-Library/hello-world.hpp]: Not sure what this header is, but you don't appear to be using it so I'd drop it. Also it should be `std::string say_hello();` I presume you were having a compiler error?

**Review by ryanmrichard — CHANGES_REQUESTED** [review]
Overall your most recent changes seem to be a step backwards. For this PR we're trying to:

1. Build your hello world library as an external project
2. Test the hello world library that you built as an external project
3. Install the hello world library

Once we have that working I think we should merge the PR. Then we can focus on adding a dependency. If I were you I'd revert to the previously reviewed commit (looks like it was 90f66ebe87635f85cea251cf4963beded3bd2a41). Then focus on removing CMakePublic.

File: CMakeLists.txt [hunk 15]
```diff
@@ -0,0 +1,19 @@
+cmake_minimum_required(VERSION 3.5.1)
+project(CMaizeSuperbuild VERSION 1.0.0)
+
+include(ExternalProject)
+
+include_directories(${CMAKE_CURRENT_SOURCE_DIR}/src)
+include_directories(${EXTERNAL_INSTALL_LOCATION}/Shared-Library)
```
Comment (ryanmrichard) [inline CMakeLists.txt]: Why are these here?

File: src/hello-world.cpp [hunk 16]
```diff
@@ -0,0 +1,8 @@
+#include <iostream>
+
+int HelloWorld (int argc, char** argv)
```
Comment (ryanmrichard) [inline src/hello-world.cpp]: This isn't the API you declared in the header file.

**Review by ryanmrichard — CHANGES_REQUESTED** [review]

File: src/hello-world.cpp (hunk 16 above)
Comment (eminokic) [reply src/hello-world.cpp]: I will go ahead and update that now.

File: CMakeLists.txt (hunk 15 above)
Comment (eminokic) [reply CMakeLists.txt]: They were from the older push I had, I kept them since I had the two folder in my project, but for the superbuild I will just use an external project add command to add the directories in from the repository.

Comment (ryanmrichard) [reply CMakeLists.txt]: Gotcha. Yeah I don't think you want these here for the superbuild.

File: Shared-Library/hello-world.hpp [hunk 17]
```diff
@@ -1,4 +1,3 @@
-#pragma once
```
Comment (ryanmrichard) [inline Shared-Library/hello-world.hpp]: You still need the `#pragma once`

File: src/hello-world.cpp [hunk 18]
```diff
@@ -1,8 +1,5 @@
+#include <cstdlib>
 #include <iostream>
```
Comment (ryanmrichard) [inline src/hello-world.cpp]: You're no longer using these headers.

File: src/hello-world.cpp [hunk 19]
```diff
@@ -0,0 +1,3 @@
+#include <string>
```
Comment (ryanmrichard) [inline src/hello-world.cpp]: It's standard C/C++ practice to include the header file defining the API, i.e., `#include "hello-world.hpp"`, which also pulls in your public dependencies (`<string>`).

File: src/hello-world.cpp [hunk 20]
```diff
@@ -0,0 +1,3 @@
+#include "../Shared-Library/hello-world.hpp"
```
Comment (ryanmrichard) [inline src/hello-world.cpp]: This should be `#include "Shared-Library/hello-world.hpp"`. Admittedly It doesn't matter so much for source files since they aren't referenced in other phases of the build, but doing so will help make sure you get the include paths right.

File: CMakeLists.txt [hunk 21]
```diff
@@ -0,0 +1,18 @@
+cmake_minimum_required(VERSION 3.5.1)
+project(CMaizeSuperbuild VERSION 1.0.0)
+
+include(ExternalProject)
+
+include(GNUInstallDirs)
+
```
Comment (ryanmrichard) [inline CMakeLists.txt:7]: You need to define a staging directory

File: CMakeLists.txt [hunk 22]
```diff
@@ -0,0 +1,18 @@
...
+    "${PROJECT_NAME}_core"
+    SOURCE_DIR "${CMAKE_CURRENT_LIST_DIR}/src"
+    CMAKE_ARGS -DCMAKE_CXX_COMPILER="${CMAKE_CXX_COMPILER}"
+               -DCMAKE_CXX_STANDARD="${CMAKE_CXX_STANDARD}"
+               -DCMAKE_CXX_EXTENSIONS="${CMAKE_CXX_EXTENSIONS}"
+               -DCMAKE_CXX_STANDARD_REQUIRED="${CMAKE_CXX_STANDARD_REQUIRED}"
+    CMAKE_CACHE_ARGS -DCMAKE_CXX_FLAGS:STRING="${CMAKE_CXX_FLAGS}"
+    BUILD_ALWAYS 1
+    INSTALL_COMMAND ""
+)
```
Comment (ryanmrichard) [inline CMakeLists.txt]: You need to install the library.

File: CMakeLists.txt [hunk 23]
```diff
@@ -0,0 +1,18 @@
+cmake_minimum_required(VERSION 3.5.1)
+project(CMaizeSuperbuild VERSION 1.0.0)
+
+include(ExternalProject)
+
+include(GNUInstallDirs)
```
Comment (ryanmrichard) [inline CMakeLists.txt:6]: This needs to move down to the `src/CMakeLists.txt` file where you actually use it.

File: CMakeLists.txt [hunk 24]
```diff
@@ -0,0 +1,18 @@
...
+
+ExternalProject_Add(
+    "${PROJECT_NAME}_core"
+    SOURCE_DIR "${CMAKE_CURRENT_LIST_DIR}/src"
+    CMAKE_ARGS -DCMAKE_CXX_COMPILER="${CMAKE_CXX_COMPILER}"
+               -DCMAKE_CXX_STANDARD="${CMAKE_CXX_STANDARD}"
+               -DCMAKE_CXX_EXTENSIONS="${CMAKE_CXX_EXTENSIONS}"
+               -DCMAKE_CXX_STANDARD_REQUIRED="${CMAKE_CXX_STANDARD_REQUIRED}"
```
Comment (ryanmrichard) [inline CMakeLists.txt]: The library should be setting the standard it needs.

File: CMakeLists.txt [hunk 25]
```diff
@@ -0,0 +1,18 @@
...
+include(ExternalProject)
+
+include(GNUInstallDirs)
+
+ExternalProject_Add(
+    "${PROJECT_NAME}_core"
+    SOURCE_DIR "${CMAKE_CURRENT_LIST_DIR}/src"
+    CMAKE_ARGS -DCMAKE_CXX_COMPILER="${CMAKE_CXX_COMPILER}"
+               -DCMAKE_CXX_STANDARD="${CMAKE_CXX_STANDARD}"
+               -DCMAKE_CXX_EXTENSIONS="${CMAKE_CXX_EXTENSIONS}"
```
Comment (ryanmrichard) [inline CMakeLists.txt]: Do you really need to set this? I feel like CMake takes care of this automatically.

File: src/CMakeLists.txt [hunk 26]
```diff
@@ -0,0 +1,28 @@
+cmake_minimum_required(VERSION 3.5.1)
```
Comment (ryanmrichard) [inline src/CMakeLists.txt]: You don't need this here. The top-level file should take care of this for the whole project.

File: src/CMakeLists.txt [hunk 27]
```diff
@@ -0,0 +1,28 @@
...
+
+project(CMaizeSuperbuild LANGUAGES CXX)
+
+add_library(${PROJECT_NAME} SHARED hello-world.cpp)
```
Comment (ryanmrichard) [inline src/CMakeLists.txt]: Don't hard code the library to a shared library.

File: src/CMakeLists.txt [hunk 28]
```diff
@@ -0,0 +1,28 @@
+cmake_minimum_required(VERSION 3.5.1)
+
+project(CMaizeSuperbuild LANGUAGES CXX)
```
Comment (ryanmrichard) [inline src/CMakeLists.txt]: This project should probably be named "HelloWorld" or something. The superbuild part is really in the top-level lists.

**Review by ryanmrichard — CHANGES_REQUESTED** [review]
I'd recommend putting the source file in `src` and header file in `include` more akin to a normal GNU structure.

File: src/hello-world.cpp (hunk 20 above)
Comment (eminokic) [reply src/hello-world.cpp]: I was getting an error building it and locating the hpp file unless it was specified as ../

File: src/CMakeLists.txt (hunk 28 above)
Comment (eminokic) [reply src/CMakeLists.txt]: I was reading that when doing multiple level CMakeLists you should use the same project name and have it drawn as a prefix build. I can remove it still if you believe that is better for the super-build.

File: CMakeLists.txt (hunk 25 above)
Comment (eminokic) [reply CMakeLists.txt]: All the super build models I have seen use this structure, but I can look into this more.

File: src/hello-world.cpp (hunk 20 above)
Comment (ryanmrichard) [reply src/hello-world.cpp]: That's because your include path isn't setup correctly.

File: CMakeLists.txt (hunk 25 above)
Comment (ryanmrichard) [reply CMakeLists.txt]: I'm referring to setting `CMAKE_CXX_EXTENSIONS`. I don't think you need to mess with it.

File: src/CMakeLists.txt (hunk 28 above)
Comment (ryanmrichard) [reply src/CMakeLists.txt]: I'd be curious to hear their reasoning for this (if they give it), but that sounds like a bad idea to me. The project name shows up in a couple places, notably in the logs. If you use the same name it's harder to distinguish between the driver build (the top-level) and the subbuilds. That's why I vote for different (and more descriptive) names.

Comment (eminokic) [reply src/CMakeLists.txt]: You're right, I have changed the name till I collect more reasoning as to if it is beneficial or not to have one project name.

File: CMakeLists.txt (hunk 25 above)
Comment (eminokic) [reply CMakeLists.txt]: Sounds good, should me removed now.

File: CMakeLists.txt (hunk 24 above)
Comment (eminokic) [reply CMakeLists.txt]: So is this an unnecessary line of code as well similar to extensions?

Comment (ryanmrichard) [reply CMakeLists.txt]: yes

Comment (ryanmrichard) [reply CMakeLists.txt]: to clarify, passing `CMAKE_CXX_STANDARD` is okay for now as that's something user could have reasonably set, but I wouldn't expect them to pass `CMAKE_CXX_EXTENSIONS` or `CMAKE_CXX_STANDARD_REQUIRED` (nor should they)

File: CMakeLists.txt [hunk 29]
```diff
@@ -0,0 +1,22 @@
+cmake_minimum_required(VERSION 3.5.1)
+project(CMaizeSuperbuild VERSION 1.0.0)
+
+include(ExternalProject)
+
+include(GNUInstallDirs)
+
+set(
+        CMAKE_INSTALL_PREFIX "${CMAKE_BINARY_DIR}/stage"
```
Comment (ryanmrichard) [inline CMakeLists.txt]: As a general rule of thumb you never want to call `set` with a variable that starts with `CMAKE`. Those variables are how users set options. If you call `set` with one of them you clobber the user's setting. In this case I recommend you call the staging directory something like `STAGING_DIRECTORY`.

**Review by ryanmrichard — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 29 above)
Comment (eminokic) [reply CMakeLists.txt]: Would you recommend it be 

CMAKE_INSTALL_PREFIX "${STAGING_DIRECTORY}/stage"
CMAKE_INSTALL_PREFIX "STAGING_DIRECTORY/stage"

or remove the CMAKE_INSTALL_PREFIX with STAGING_DIRECTORY

Comment (ryanmrichard) [reply CMakeLists.txt]: `set(STAGING_DIRECTORY ${CMAKE_BINARY_DIR}/stage/)` then in `ExternalProject_Add` pass the staging directory as the install directory like  `CMAKE_ARGS -DCMAKE_INSTALL_PREFIX="${STAGING_DIRECTORY}"`

File: src/hello-world.cpp (hunk 20 above)
Comment (eminokic) [reply src/hello-world.cpp]: Would that just be a my computer thing, or is the project path not setup correctly?

File: CMakeLists.txt (hunk 21 above)
Comment (ryanmrichard) [reply CMakeLists.txt:7]: Don't bother caching the value of the staging directory unless you have a reason to.

File: CMakeLists.txt [hunk 30]
```diff
@@ -0,0 +1,23 @@
...
+        STAGING_DIRECTORY "${CMAKE_BINARY_DIR}/stage/"
+        CACHE STRING "Stage Install Location"
+)
+
+ExternalProject_Add(
+    "${PROJECT_NAME}_core"
+    SOURCE_DIR "${CMAKE_CURRENT_LIST_DIR}/src"
+    CMAKE_ARGS -DCMAKE_CXX_COMPILER="${CMAKE_CXX_COMPILER}"
+               -DCMAKE_CXX_STANDARD="${CMAKE_CXX_STANDARD}"
+               -DCMAKE_INSTALL_PREFIX:PATH="${CMAKE_INSTALL_PREFIX}"
```
Comment (ryanmrichard) [inline CMakeLists.txt]: You're setting `CMAKE_INSTALL_PREFIX` twice. This line needs to go away.

File: src/hello-world.cpp (hunk 20 above)
Comment (ryanmrichard) [reply src/hello-world.cpp]: I'd argue it's a problem with your project structure, but you could equivalently argue it's a problem with your `src/CMakeLists.txt` setting the wrong include paths. Right now you have your includes in `Shared-Library`, but your `src/CMakeLists.txt` claims they're in `include`; the latter is the more canonical place to put them.

**Review by ryanmrichard — CHANGES_REQUESTED** [review]

File: hello_world/src/hello-world.cpp [hunk 31]
```diff
@@ -1,3 +1,3 @@
-#include "hello-world.hpp"
+#include "../include/hello-world.hpp"
```
Comment (ryanmrichard) [inline hello_world/src/hello-world.cpp]: Don't use includes relative the source file; includes should be relative to the include path. It should just be `#include "hello-world.hpp"`

File: CMakeLists.txt [hunk 32]
```diff
@@ -0,0 +1,25 @@
...
+    SOURCE_DIR "${CMAKE_CURRENT_LIST_DIR}/hello_world/src"
+    CMAKE_ARGS -DCMAKE_CXX_COMPILER="${CMAKE_CXX_COMPILER}"
+               -DCMAKE_CXX_STANDARD="${CMAKE_CXX_STANDARD}"
+               -DCMAKE_INSTALL_PREFIX="${STAGING_DIRECTORY}"
+    CMAKE_CACHE_ARGS -DCMAKE_CXX_FLAGS:STRING="${CMAKE_CXX_FLAGS}"
+    BUILD_ALWAYS 1
+    INSTALL_COMMAND ""
+)
+
+ExternalProject_Get_Property(${PROJECT_NAME}_core install_dir)
```
Comment (ryanmrichard) [inline CMakeLists.txt]: I  presume you are trying to install? You already know where it is supposed to be installed to (`CMAKE_INSTALL_PREFIX`). For a superbuild, installing should just be a matter of copying the staging directory to the final destination.

File: CMakeLists.txt [hunk 33]
```diff
@@ -0,0 +1,25 @@
...
+
+include_directories(${CMAKE_BINARY_DIR}/stage)
```
Comment (ryanmrichard) [inline CMakeLists.txt]: Why is this here?

Comment (eminokic) [reply CMakeLists.txt]: I thought in order to make use of the staging directory then it would be required to include it. I will look into refactoring that.

File: CMakeLists.txt (hunk 32 above)
Comment (eminokic) [reply CMakeLists.txt]: Is this unnecessary then?

File: CMakeLists.txt (hunk 33 above)
Comment (ryanmrichard) [reply CMakeLists.txt]: make use of it how? This is telling CMake where to look for header files when it goes to build a C++ project, but you've already built all the C++ projects at this point.

File: CMakeLists.txt (hunk 32 above)
Comment (ryanmrichard) [reply CMakeLists.txt]: yes

File: CMakeLists.txt [hunk 34]
```diff
@@ -18,4 +22,10 @@ ExternalProject_Add(
     CMAKE_CACHE_ARGS -DCMAKE_CXX_FLAGS:STRING="${CMAKE_CXX_FLAGS}"
     BUILD_ALWAYS 1
     INSTALL_COMMAND ""
+)
+
+add_executable(
```
Comment (ryanmrichard) [inline CMakeLists.txt]: If you're going to make an executable it's got to be done as an external project also. I wouldn't worry about making one until you've got the installation working. Once you get the installation working let's merge this PR, then we can worry about adding dependencies and/or an executable.

File: hello_world/src/CMakeLists.txt [hunk 35]
```diff
@@ -5,7 +5,7 @@ include(ExternalProject)
 include(GNUInstallDirs)
 
 set_target_properties(${PROJECT_NAME} PROPERTIES PUBLIC_HEADER
-        ${CMAKE_CURRENT_SOURCE_DIR}/include/hello-world.hpp
+        ../include/hello-world.hpp
```
Comment (ryanmrichard) [inline hello_world/src/CMakeLists.txt]: You should always use absolute paths in CMake files. Also why did you move this file?

Comment (eminokic) [reply hello_world/src/CMakeLists.txt]: I am not sure why it moved let me check and fix that.

File: CMakeLists.txt [hunk 36]
```diff
@@ -0,0 +1,27 @@
...
+
+ExternalProject_Add(
+    "${PROJECT_NAME}_core"
+    SOURCE_DIR "${CMAKE_CURRENT_LIST_DIR}/hello_world"
+    CMAKE_ARGS -DCMAKE_CXX_COMPILER="${CMAKE_CXX_COMPILER}"
+               -DCMAKE_CXX_STANDARD="${CMAKE_CXX_STANDARD}"
+               -DCMAKE_INSTALL_PREFIX="${STAGING_DIRECTORY}"
+    CMAKE_CACHE_ARGS -DCMAKE_CXX_FLAGS:STRING="${CMAKE_CXX_FLAGS}"
+    BUILD_ALWAYS 1
+        DESTDIR=${CMAKE_BINARY_DIR}/stage
```
Comment (ryanmrichard) [inline CMakeLists.txt]: This is just `STAGING_DIRECTORY`; also does `ExternalProject_Add` recognize `DESTDIR`?

File: hello_world/CMakeLists.txt [hunk 37]
```diff
@@ -0,0 +1,31 @@
+cmake_minimum_required(VERSION ${CMAKE_VERSION})
+project(HelloWorld LANGUAGES CXX)
+
+include(ExternalProject)
```
Comment (ryanmrichard) [inline hello_world/CMakeLists.txt]: You don't need `ExternalProject` here

File: hello_world/CMakeLists.txt [hunk 38]
```diff
@@ -0,0 +1,31 @@
...
+include(GNUInstallDirs)
+
+add_subdirectory(src)
```
Comment (ryanmrichard) [inline hello_world/CMakeLists.txt]: don't `add_subdirectory` `src` and `include`, just define the library in this file.

**Comment by ryanmrichard** [comment]
I finished this off in PR #2. The next step is to add CMakePublic as a dependency to the super build.
